### PR TITLE
Coot 13, Change `arma::fill` to `internal_compat` to avoid confusion with user declarations

### PR DESCRIPTION
**Pull request by shrit**

### CODE REVIEW

File: src/mlpack/core/util/using.hpp [hunk 1]
```diff
@@ -71,6 +72,49 @@ namespace mlpack {
 
 #endif
 
+  namespace internal_compact {
+
+    namespace fill {
+
+      #ifdef MLPACK_HAS_COOT
+        struct fill_none  : public arma::fill::fill_class<
+                              arma::fill::fill_none>,
```
Comment (conradsnicta) [inline src/mlpack/core/util/using.hpp:81]: I strongly advise against directly using `arma::fill::fill_class<>`.  This is an implementation level detail that can be changed in future versions of Armadillo.

Only the following symbols are publicly documented within Armadillo docs: `arma::fill::none`, `arma::fill::zeros`, `arma::fill::ones`, `arma::fill::eye`, `arma::fill::randu`, `arma::fill::randn`, `arma::fill::value`.

Extract from the on [API stability section](https://arma.sourceforge.net/docs.html#changelog) of Armadillo documentation:  
> ...     Any functionality within Armadillo which is not explicitly described in the public API documentation is considered as internal implementation detail, and may be changed or removed without notice.

In other words, `arma::fill::fill_class<>` is off-limits.

Comment (rcurtin) [reply src/mlpack/core/util/using.hpp:81]: @conradsnicta this one is tricky. I agree and would prefer not to use internal Armadillo types, but here we want one name we can use that can be *both* `arma::fill::zeros` and `coot::fill::zeros`.  This helps us avoid the situation where mlpack developers can't use `fill::`, which would feel like an awkward restriction.  Ideally writing internal mlpack code feels as much like writing Armadillo as possible, except to keep it general we avoid the `arma::`.  Without some namespace-agnostic solution for `fill::`, we can't have any uninitialized matrices, which is sometimes desired. (i.e. avoid filling the matrix with zeros, since the next call fills it.)

Under those desiderata, the only solution I could come up with is inheriting from the base class underlying both `arma::fill::zeros` and `coot::fill::zeros`. Do you have any alternative suggestions?  I'm out today so I don't think I'll be able to play with it until Monday. Maybe there is some clever trick that C++ allows that we haven't thought of yet.

Comment (rcurtin) [reply src/mlpack/core/util/using.hpp:81]: Maybe we can cleverly use `decltype` to avoid the direct name `arma::fill::fill_class`? We would still be depending on the fact that `arma::fill::zeros` is a type and not an enum.

Comment (rcurtin) [reply src/mlpack/core/util/using.hpp:81]: :facepalm: oh, I was way overthinking it, yeah, a simple template metaprogramming approach like that would work great.  It's a couple more characters, but that's not an issue.

> PS. I'm still not convinced that abstracting away the matrix type is the right way forward, though that discussion should probably be continued in the contributors-chat repo. All of this has the smell of overengineering.

I see where you're coming from, but I also don't see another reasonable way to allow mlpack to work on matrix types that aren't supported directly by Armadillo.

**Comment by conradsnicta** [comment]
@shrit @rcurtin Followup to the earlier discussion in https://github.com/mlpack/mlpack/pull/3687#discussion_r1562001400.

Here's a possible implementation that avoids the direct use of internal functionality within Armadillo and Bandicoot.  This requires C++17 due to use of inline variables (to my current understanding of template voodoo).  Under C++14 this may cause linking errors.  Example usage is listed after the proposed code.

Feel free to copy'n'paste or adapt this into mlpack under the BSD license, though please include me as one of the authors.

```
template<typename MatType>
struct GetFillMode
  {
  };

// specialisation for Armadillo matrices
template<typename eT>
struct GetFillMode< arma::Mat<eT> >
  {
  static constexpr decltype(arma::fill::none ) none  = arma::fill::none;
  static constexpr decltype(arma::fill::zeros) zeros = arma::fill::zeros;
  static constexpr decltype(arma::fill::ones ) ones  = arma::fill::ones;
  static constexpr decltype(arma::fill::eye  ) eye   = arma::fill::eye;
  static constexpr decltype(arma::fill::randu) randu = arma::fill::randu;
  static constexpr decltype(arma::fill::randn) randn = arma::fill::randn;
  };

// TODO: to increase ease of use, include specialisations for arma::Col<eT> and arma::Row<eT>

#if defined(MLPACK_HAS_COOT)

// specialisation for Bandicoot matrices
template<typename eT>
struct GetFillMode< coot::Mat<eT> >
  {
  static constexpr decltype(coot::fill::none ) none  = coot::fill::none;
  static constexpr decltype(coot::fill::zeros) zeros = coot::fill::zeros;
  static constexpr decltype(coot::fill::ones ) ones  = coot::fill::ones;
  static constexpr decltype(coot::fill::eye  ) eye   = coot::fill::eye;
  static constexpr decltype(coot::fill::randu) randu = coot::fill::randu;
  static constexpr decltype(coot::fill::randn) randn = coot::fill::randn;
  };

#endif
```

Example usage:
```
arma::mat A(10, 10, GetFillMode<arma::mat>::randu);
```

PS. I haven't yet looked into an abstraction for `arma::fill::value(scalar)`, as that's likely to be more involved.  As a possible workaround this pattern can be used instead:
```
MatType A(10, 10, GetFillMode<MatType>::none);
A.fill(scalar); 
```

**Comment by rcurtin** [comment]
Thanks!  That will definitely work.

I wondered about the code you wrote---instead of naming the type of `arma::fill::zeros`, it just uses `decltype()` to avoid doing that.  One could actually do the same thing for the original strategy:

```
    namespace fill {
      // used when both Armadillo and Bandicoot are available
      class fill_zeros : public decltype(arma::fill::zeros), public decltype(coot::fill::zeros) { };

      static constexpr fill_zeros zeros;
    };
```

How do you feel about that implementation---is that using Armadillo internals in an undesired way?

I'm ambivalent about which to use, personally, I suspect the template metaprogramming selector approach is nicer to work with than the multiple inheritance approach.

> PS. I haven't yet looked into an abstraction for arma::fill::value(scalar), as that's likely to be more involved. As a possible workaround this pattern can be used instead:

Yeah, I see, it is a little trickier but I think it could be done.  Basically the correct function would need to be forwarded based on the given matrix type.  I think that Bandicoot doesn't current implement `fill::value` though, and we also don't use it internally, so not a huge issue to get that right now.

**Comment by conradsnicta** [comment]
This doesn't directly touch Armadillo internals, so on the surface this looks fine.  I think that when Armadillo internals are updated, mlpack-based code would only need a recompilation.  Since mlpack is now header only, and almost all of Armadillo is in headers, there shouldn't be any breakage.

I recommend double checking that the multiple inheritance approach actually works.  In other words, check that `fill::none`, `fill::zeros`, `fill::ones`, `fill::randu` etc still fill the matrices with correct values.

**Comment by shrit** [comment]
@rcurtin maybe it is the time for us to jump to C++17 to make this compile on the CI, because it is passing with no issue on my end
1. Currently we are close to May, which is the middle of the year.
2. We will need sometime to adjust and remove all the backporting that we have added

**Comment by shrit** [comment]
As discussed internally, I will get back to this after we move to C++17

File: src/mlpack/core/util/using.hpp [hunk 2]
```diff
@@ -71,6 +73,37 @@ namespace mlpack {
...
+
+        struct fill_randu : public decltype(arma::fill::randu),
+                            public decltype(coot::fill::randu) { };
+
+      #else
+        struct fill_none  : public decltype(arma::fill::none) { };
+        struct fill_zeros : public decltype(arma::fill::zeros) { };
+        struct fill_ones  : public decltype(arma::fill::ones) { };
+        struct fill_randu : public decltype(arma::fill::randu) { };
+      #endif
```
Comment (rcurtin) [inline src/mlpack/core/util/using.hpp:98]: I think that Conrad's suggestion is a safer way to go here:

```c++
// By default, assume that we are using an Armadillo object.
template<typename MatType>
struct GetFillType
{
  static constexpr decltype(arma::fill::none) none = arma::fill::none;
  static constexpr decltype(arma::fill::zeros) zeros = arma::fill::zeros;
  static constexpr decltype(arma::fill::ones) ones = arma::fill::ones;
  static constexpr decltype(arma::fill::randu) randu = arma::fill::randu;
  static constexpr decltype(arma::fill::randn) randn = arma::fill::randn;
};

#ifdef MLPACK_HAS_COOT
// If the matrix type is a Bandicoot type, use Bandicoot fill objects instead.
template<typename MatType,
         typename = typename std::enable_if<is_coot_type<MatType>::value>::type*>
struct GetFillType
{
  static constexpr decltype(coot::fill::none) none = coot::fill::none;
  static constexpr decltype(coot::fill::zeros) zeros = coot::fill::zeros;
  static constexpr decltype(coot::fill::ones) ones = coot::fill::ones;
  static constexpr decltype(coot::fill::randu) randu = coot::fill::randu;
  static constexpr decltype(coot::fill::randn) randn = coot::fill::randn;
};
#endif
```

I haven't tried compiling that, but the general idea should be adaptable into something that works. :+1:

**Comment by rcurtin** [comment]
Have you tried using an older version of gcc to reproduce locally?  We can't expect everyone to have the newest version of the compiler.  mlpack should work with any C++17 compiler, including older versions of g++.

File: src/mlpack/core/util/using.hpp [hunk 3]
```diff
@@ -75,6 +77,61 @@ namespace mlpack {
...
+    // If the matrix type is a Bandicoot type, use Bandicoot fill objects instead.
+    template<typename MatType,
+             typename = typename std::enable_if<is_coot_type<MatType>::value>::type*>
+    struct GetFillType
+    {
+      static constexpr decltype(coot::fill::none) none = coot::fill::none;
+      static constexpr decltype(coot::fill::zeros) zeros = coot::fill::zeros;
+      static constexpr decltype(coot::fill::ones) ones = coot::fill::ones;
+      static constexpr decltype(coot::fill::randu) randu = coot::fill::randu;
+      static constexpr decltype(coot::fill::randn) randn = coot::fill::randn;
```
Comment (rcurtin) [inline src/mlpack/core/util/using.hpp:102]:
```suggestion
      static constexpr decltype(coot::fill::none) none;
      static constexpr decltype(coot::fill::zeros) zeros;
      static constexpr decltype(coot::fill::ones) ones;
      static constexpr decltype(coot::fill::randu) randu;
      static constexpr decltype(coot::fill::randn) randn;
```
I think that this should fix the compilation errors---since the `fill::` types depend on types and not actual instances of those types, I don't think it's necessary to actually make the objects explicitly equal to the ones held internally in Armadillo or Bandicoot.

File: src/mlpack/core/util/using.hpp [hunk 4]
```diff
@@ -75,6 +77,61 @@ namespace mlpack {
...
+    {
+      static const decltype(coot::fill::none) none;
+      static const decltype(coot::fill::zeros) zeros;
+      static const decltype(coot::fill::ones) ones;
+      static const decltype(coot::fill::randu) randu;
+      static const decltype(coot::fill::randn) randn;
+    };
+  #endif
+
+#else
```
Comment (rcurtin) [inline src/mlpack/core/util/using.hpp:106]: There's no difference in the `#else` here; but if I remember right, newer Armadillo versions use `constexpr` and older ones use `const`.  Can you make the newer version use `constexpr`?

File: src/mlpack/tests/split_data_test.cpp [hunk 5]
```diff
@@ -94,8 +94,7 @@ void CheckDuplication(const Row<size_t>& trainLabels,
 TEST_CASE("SplitShuffleDataResultMat", "[SplitDataTest]")
 {
   mat input(2, 10);
-  size_t count = 0; // Counter for unique sequential values.
-  input.imbue([&count] () { return ++count; });
+  input = linspace(0, input.n_elem - 1);
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:97]: Don't you need to specify `N` here? https://arma.sourceforge.net/docs.html#linspace

linspace gives a vector though, so if you want this to be a matrix, maybe you need to do something like... `mat input = reshape(linspace<mat>(0, 19, 20), 2, 10);` (you should double-check that, I am not 100% sure)

**Review by rcurtin — COMMENTED** [review]
Looks like just a couple tests are failing but I think overall it works?

File: src/mlpack/core/util/using.hpp (hunk 4 above)
Comment (shrit) [reply src/mlpack/core/util/using.hpp:106]: oh, I just removed all of that, I do not think I have pushed the modification yet
I only left one case, for all, I think when we make the base version is arma 12, then we can use `constexpr`

Comment (rcurtin) [reply src/mlpack/core/util/using.hpp:106]: If it works with `const`, that's fine with me, although it would be worth adding a comment to indicate the minimum required Armadillo version to switch to `constexpr` (I think that is in some other comment in this PR somewhere). :+1:

File: src/mlpack/tests/split_data_test.cpp (hunk 5 above)
Comment (shrit) [reply src/mlpack/tests/split_data_test.cpp:97]: Let us if the tests pass :smiley:

**Comment by shrit** [comment]
This is ready for merging, the tests are passing locally, will see on the CI :smiley:

File: src/mlpack/methods/ann/loss_functions/hinge_loss_impl.hpp [hunk 6]
```diff
@@ -31,9 +31,9 @@ typename MatType::elem_type HingeLossType<MatType>::Forward(
     const MatType& target)
 {
   MatType temp = target - (target == 0);
-  MatType temp_zeros(size(target), arma::fill::zeros);
+  MatType tempZeros(size(target));
```
Comment (rcurtin) [inline src/mlpack/methods/ann/loss_functions/hinge_loss_impl.hpp:34]: Nice find :)

File: src/mlpack/methods/lars/lars_impl.hpp [hunk 7]
```diff
@@ -573,9 +573,10 @@ LARS<ModelMatType>::Train(const MatType& matX,
   isIgnored.resize(dataRef.n_cols, false);
 
   // Initialize yHat and beta.
-  arma::Col<ElemType> beta(dataRef.n_cols, arma::fill::zeros);
-  arma::Col<ElemType> yHat(dataRef.n_rows, arma::fill::zeros);
-  arma::Col<ElemType> yHatDirection(dataRef.n_rows, arma::fill::none);
+  arma::Col<ElemType> beta(dataRef.n_cols);
+  arma::Col<ElemType> yHat(dataRef.n_rows);
+  arma::Col<ElemType> yHatDirection(dataRef.n_rows,
+      arma::fill::none);
```
Comment (rcurtin) [inline src/mlpack/methods/lars/lars_impl.hpp:579]:
```suggestion
  arma::Col<ElemType> yHatDirection(dataRef.n_rows, arma::fill::none);
```

File: src/mlpack/methods/lmnn/constraints_impl.hpp [hunk 8]
```diff
@@ -405,7 +405,7 @@ void Constraints<MatType, LabelsType, DistanceType>::Triplets(
   UMatType targetNeighbors(k, dataset.n_cols);;
   TargetNeighbors(targetNeighbors, dataset, labels, norms);
 
-  outputMatrix = UMatType(3, k * k * N , arma::fill::zeros);
+  outputMatrix = UMatType(3, k * k * N );
```
Comment (rcurtin) [inline src/mlpack/methods/lmnn/constraints_impl.hpp:408]:
```suggestion
  outputMatrix = UMatType(3, k * k * N);
```

File: src/mlpack/methods/lsh/lsh_search_impl.hpp [hunk 9]
```diff
@@ -299,7 +299,8 @@ void LSHSearch<SortPolicy, MatType>::Train(MatType referenceSet,
 
   // Now, using the hash vectors for each table, count the number of rows we
   // have in the second hash table.
-  arma::Row<size_t> secondHashBinCounts(secondHashSize, arma::fill::zeros);
+  arma::Row<size_t> secondHashBinCounts(secondHashSize,
+      arma::fill::zeros);
```
Comment (rcurtin) [inline src/mlpack/methods/lsh/lsh_search_impl.hpp:303]:
```suggestion
  arma::Row<size_t> secondHashBinCounts(secondHashSize);
```
Seems like no need for the `zeros`?

File: src/mlpack/tests/split_data_test.cpp [hunk 10]
```diff
@@ -362,10 +357,9 @@ TEST_CASE("SplitDataResultField", "[SplitDataTest]")
   mat matA(2, 10);
   mat matB(2, 10);
 
-  size_t count = 0; // Counter for unique sequential values.
-  matA.imbue([&count]() { return ++count; });
-  matB.imbue([&count]() { return ++count; });
-
+  matA = linspace(0, matA.n_elem - 1);
+  matA = linspace(matA.n_elem, matA.n_elem + matB.n_elem - 1);
```
Comment (rcurtin) [inline src/mlpack/tests/split_data_test.cpp:361]: Should this be setting `matB` instead of `matA`?

```suggestion
  matB = linspace(matA.n_elem, matA.n_elem + matB.n_elem - 1);
```

**Review by rcurtin — APPROVED** [review]
Everything looks good to me.  There is one comment in `split_data_test.cpp` that I think is a bug, but all my other comments are tiny little fixes (as usual :)).

File: src/mlpack/tests/split_data_test.cpp (hunk 10 above)
Comment (shrit) [reply src/mlpack/tests/split_data_test.cpp]: Yes you are right, probably a regexp error, funny enough, it was not caught by  tests
